pandas/core/categorical.py

```diff
@@ -1364,6 +1364,24 @@ def sort_values(self, inplace=False, ascending=True, na_position='last'):
             return self._constructor(values=codes, categories=self.categories,
                                      ordered=self.ordered, fastpath=True)
 
+    def _values_for_rank(self):
+        """
+        For correctly ranking ordered categorical data. See GH#15420
+
+        Ordered categorical data should be ranked on the basis of
+        codes with -1 translated to NaN.
+
+        Returns
+        -------
+        numpy array
+
+        """
+        values = self._codes.astype('float64')
+        if self._ordered:
+            na_mask = (values == -1)
+            values[na_mask] = np.nan
+        return values
```

Comment:
pls do it like I shows. this MUST be in here and not in external scope (e.g. algos)